Add initial version of MXNet bindings and wrapper library

joseph-wakeling-sociomantic opened the pull request:
This patch provides the first public release of the `dmxnet` library code, including both the raw C bindings (`mxnet.c.c_api`) and higher level D wrapper classes for more idiomatic use.

Submodules have been added for `makd` and `ocean` together with basic build scripts.

* submodules/makd ()...v1.9.1(8711da2) (1 commits)
  > Merge tag 'v1.8.4' into v1.9.x

* submodules/ocean ()...v3.3.1(a991c41) (1 commits)
  > Merge tag v2.9.1 into v3.3.x

joseph-wakeling-sociomantic committed:
Add initial version of MXNet bindings and wrapper library

This patch provides the first public release of the `dmxnet` library
code, including both the raw C bindings (`mxnet.c.c_api`) and higher
level D wrapper classes for more idiomatic use.

Submodules have been added for `makd` and `ocean` together with basic
build scripts.  A `.gitignore` file has also been added to cover the
common cases anticipated while building or contributing to the code,
plus a `.neptune.yml` to assist tracking of dependencies.

The top-level `LICENSE.txt` contains the Boost License Version 1.0
applied to the wrapper library, while a copy of the Apache License
Version 2.0 accompanies the C API bindings.

* submodules/makd ()...v1.9.1(8711da2) (1 commits)
  > Merge tag 'v1.8.4' into v1.9.x

* submodules/ocean ()...v3.3.1(a991c41) (1 commits)
  > Merge tag v2.9.1 into v3.3.x

joseph-wakeling-sociomantic commented:
I'll follow up on this with travis setup, Docker images, etc.  Note that I'm deliberately excluding our test suite for the time being as this will add some complications for Docker etc.

joseph-wakeling-sociomantic commented on src/mxnet/MXNet.d:
```diff
@@ -0,0 +1,122 @@
+    // a label is predicted for some unseen example x
+    // by returning the index for which x*W' + b is maximal
+    scope feature_matrix = new Variable("X");
+    scope(exit) feature_matrix.freeHandle();
```
Minor style: should probably take the opportunity to correct all these to `scope (exit)`.

jens-mueller-sociomantic replied:
I believe there is nothing in the style guide saying that you should put a space here. The style only requires [a space for function definitions](https://github.com/sociomantic/backend/blob/master/doc/coding/d-style-guide.rst#spaces). It seems you want a space after each keyword. Because we do that for `version`, `debug`, etc. But for some reason we don't do for `assert`. I find the rule for function definitions convincing to allow textual search. But I see no argument for `scope()`, `version()`, etc.

joseph-wakeling-sociomantic replied:
It looks like it's a de facto standard in ocean:
https://github.com/sociomantic-tsunami/ocean/search?utf8=%E2%9C%93&q=scope%28exit%29&type=

BTW regarding `assert`: some people do prefer to place a space after it, but it seems to be more consistently treated as if it's a function call.  I think the distinction here is not so much about keywords versus non-keywords, but control flow versus calls.

jens-mueller-sociomantic commented on Build.mak:
```diff
@@ -0,0 +1,15 @@
+export ASSERT_ON_STOMPING_PREVENTION=1
+
+override DFLAGS += -w
+override LDFLAGS += -lopenblas -lmxnet
```
I think we don't need to link `openblas` here. It should be already included in `libmxnet`.

joseph-wakeling-sociomantic commented:
Updated to fix `scope (exit)` style and added a `.gitignore`.  I've also removed `libopenblas` from the `LDFLAGS`.

joseph-wakeling-sociomantic commented:
I've added a `.neptune.yml` file to the commit, and added some notes on this and on the license files to the commit message.

joseph-wakeling-sociomantic committed:
Add basic CI setup

This patch adds dockerfile definitions for a clean build environment and
a minimal setup to run tests in Travis CI, using the `beaver` library to
provide the required functionality.

* submodules/beaver ()...v0.1.2(f60fd6c) (1 commits)
  > docker: Fix locale to C.UTF-8

joseph-wakeling-sociomantic commented on Build.mak:
```diff
@@ -0,0 +1,15 @@
+	override DFLAGS += -v2 -v2=-static-arr-params -v2=-volatile
+endif
+
+test: override LDFLAGS += -lz
```
This `override` is also unnecessary for the initial commit.

jens-mueller-sociomantic replied:
Why is that?

joseph-wakeling-sociomantic replied:
@jens-mueller-sociomantic I suspect it's not used outside the test suite.

joseph-wakeling-sociomantic commented:
Removed the unnecessary `LDFLAGS` override from the `test` build target, and added a patch with WiP of CI setup (it contains just base xenial Dockerfile and dockerfile build script for now, I'll follow up with `.travis.yml` later today).

joseph-wakeling-sociomantic commented:
I've added Travis details to the CI patch.  Note that I've tweaked the mxnet repo to use `prerelease` as a short-term test measure: this MUST be returned to `release` before this PR is merged.

joseph-wakeling-sociomantic commented on docker/build:
```diff
@@ -0,0 +1,19 @@
+echo '# sociomantic-tsunami repos' \
+    > /etc/apt/sources.list.d/sociomantic-tsunami.list
+dist="$(lsb_release -cs)"
+echo "deb https://dl.bintray.com/sociomantic-tsunami/mxnet $dist prerelease"
```
This line MUST be `release` (not `preprelease`) before this PR can be merged.  Note that this will require making a first stable release of our `libmxnet` package.

leandro-lucarella-sociomantic commented:
You should probably base the travis stuff in beaver and cachalot v2 images by now, look at [ebtree](https://github.com/sociomantic-tsunami/ebtree/blob/sociomantic/.travis.yml) as an example.

joseph-wakeling-sociomantic commented:
@leandro-lucarella-sociomantic yea, I wondered about that.  I'll take a look and fix things up accordingly: just wanted to do the minimum in order to verify that we can build.

joseph-wakeling-sociomantic commented:
Squashed and rebased with the missing `travis-script.sh` included.

joseph-wakeling-sociomantic commented:
OK, nominal success.  I agree @leandro-lucarella-sociomantic that this should switch to using beaver before we merge, but I'll make a couple of tweaks `libmxnet` side and finalize the package there before coming back to this.

joseph-wakeling-sociomantic commented on .gitignore:
```diff
@@ -0,0 +1,5 @@
+build-d[12]/
+Config.local.mak
+# Coverage files are generated by default
+*.lst
```
I just copied this file blindly from ocean TBH ;-)  I assumed that it was there for a reason (e.g. to support code coverage analysis).

joseph-wakeling-sociomantic commented:
Updated to use beaver instead of manually-written CI scripts.

joseph-wakeling-sociomantic commented:
OK, this is weird: the build is failing with a message about `ci/travis-install.sh` being found, but that should no longer be needed anyway (and is no longer referenced anywhere in `.travis.yml`).

joseph-wakeling-sociomantic commented:
A fresh force-push seems to have fixed the build.  Force-pushing again with a small fix to make sure that (unit)tests actually get run this time:

```diff
diff --git a/.travis.yml b/.travis.yml
index 6ca0719..c9b6246 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -25,4 +25,4 @@ env:
 
 install: beaver install
 
-script: beaver make all
+script: beaver make test all
```

joseph-wakeling-sociomantic commented:
OK, this looks good test-wise.  Just needs a last glance over, and then removal of the `[DO NOT MERGE]` patch (which basically means we need to get a released `libmxnet` package before we can move further on this PR).

joseph-wakeling-sociomantic commented:
`[DO NOT MERGE]` patch has been dropped now we have a first release of the deb package.  Let's see if this works ...

joseph-wakeling-sociomantic commented:
OK, all good.  Can we have a last round of review before finalizing this?

joseph-wakeling-sociomantic commented:
Updated to remove unnecessary entries from `.gitignore`:

```diff
diff --git a/.gitignore b/.gitignore
index f3ba80f..2569878 100644
--- a/.gitignore
+++ b/.gitignore
@@ -1,5 +1,2 @@
-build/
-build-d[12]/
+build*/
 Config.local.mak
-# Coverage files are generated by default
-*.lst
```

joseph-wakeling-sociomantic commented:
OK, let's take this in.  I'll try to follow up with test-suite in the next PR.
